File: backEnd-netcore/src/lunch.Repositories/Impl/Security/UserRepository.cs
using System;
using lunch.Domain.Security;
using lunch.Repositories.Security;
using Microsoft.EntityFrameworkCore;
using System.Threading.Tasks;

namespace lunch.Repositories.Impl.Security
{
    internal class UserRepository : RepositoryBase<User, int>, IUserRepository
    {
        public UserRepository(ApplicationDbContext dbContext)
            : base(dbContext)
        {
        }

        public virtual Task<User> FindById(int id)
        {
            return Set.SingleOrDefaultAsync(x => x.Id == id);
        }

        public Task<User> FindByEmail(string email)
        {
            return Set.SingleOrDefaultAsync(x => x.Email == email);
        }
    }
}

File: backEnd-go/repository/security/userRepository.go
package security

import (
	"database/sql"

	"github.com/bcrusu/lunch/backEnd-go/domain/security"
	"github.com/bcrusu/lunch/backEnd-go/repository/internal"
)

type UserRepository interface {
	Add(user security.User) int
	Delete(user security.User)
	FindByID(id int) *security.User
	FindByEmail(email string) *security.User
}

type userRepository struct {
	db *sql.DB
}

func NewUserRepository(db *sql.DB) UserRepository {
	result := new(userRepository)
	result.db = db
	return result
}

func (r *userRepository) Add(user security.User) int {
	return -1 //TODO
}

func (r *userRepository) Delete(user security.User) {
	//TODO
}

func (r *userRepository) FindByID(id int) *security.User {
	entity := internal.RunQueryRow(r.db, userFactory, selectUserBase+" WHERE u.[Id] = ?1", id)
	return entity.(*security.User)
}

func (r *userRepository) FindByEmail(email string) *security.User {
	entity := internal.RunQueryRow(r.db, userFactory, selectUserBase+" WHERE u.[Email] = ?1", email)
	return entity.(*security.User)
}

func userFactory() (interface{}, []interface{}) {
	instance := security.User{}
	scanDest := []interface{}{&instance.ID,
		&instance.Description,
		&instance.DisplayName,
		&instance.Email,
		&instance.ExternalID,
		&instance.FirstName,
		&instance.LastName,
		&instance.PictureURL,
		&instance.Type}

	return &instance, scanDest
}

const selectUserBase = `SELECT [Id],
    [Description], 
	[DisplayName],
	[Email],
	[ExternalId],
	[FirstName],
	[LastName],
	[PictureUrl],
	[Type] FROM [Users] u`
